pallets/asset-tx-payment/src/lib.rs

```diff
@@ -0,0 +1,1010 @@
+//!   inclusion_fee = base_fee + length_fee + [targeted_fee_adjustment * weight_fee];
+//!   final_fee = inclusion_fee + tip;
+//!   ```
+//!
+//!   - `targeted_fee_adjustment`: This is a multiplier that can tune the final fee based on
+//! 	the congestion of the network.
+//!
+//! Additionally, this module allows one to configure:
+//!   - The mapping between one unit of weight to one unit of fee via [`Config::WeightToFee`].
+//!   - A means of updating the fee for the next block, via defining a multiplier, based on the
+//!     final state of the chain at the end of the previous block. This can be configured via
+//!     [`Config::FeeMultiplierUpdate`]
+//!   - How the fees are paid via [`Config::OnChargeTransaction`].
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use sp_std::prelude::*;
+use codec::{Encode, Decode, EncodeLike};
+use frame_support::{
+	decl_storage, decl_module,
+	DefaultNoBound,
+	traits::Get,
+	weights::{
+		Weight, DispatchInfo, PostDispatchInfo, GetDispatchInfo, Pays, WeightToFeePolynomial,
+		WeightToFeeCoefficient, DispatchClass,
+	},
+	dispatch::DispatchResult,
+};
+use sp_runtime::{
+	FixedU128, FixedPointNumber, FixedPointOperand, Perquintill, RuntimeDebug,
+	transaction_validity::{
+		InvalidTransaction, TransactionPriority, ValidTransaction, TransactionValidityError, TransactionValidity,
+	},
+	traits::{
+		Saturating, SignedExtension, SaturatedConversion, Convert, Dispatchable,
+		DispatchInfoOf, PostDispatchInfoOf, Zero, One,
+	},
+};
+use pallet_balances::NegativeImbalance;
+use pallet_transaction_payment::OnChargeTransaction;
+use frame_support::traits::tokens::{fungibles::{Balanced, Inspect, CreditOf}, WithdrawConsequence};
+
+// mod payment;
+// mod types;
+
+// pub use payment::*;
+// pub use types::{InclusionFee, FeeDetails, RuntimeDispatchInfo};
+
+type BalanceOf<T> = <<T as pallet_transaction_payment::Config>::OnChargeTransaction as OnChargeTransaction<T>>::Balance;
+type AssetBalanceOf<T> = <<T as Config>::Fungibles as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+type AssetIdOf<T> = <<T as Config>::Fungibles as Inspect<<T as frame_system::Config>::AccountId>>::AssetId;
+type LiquidityInfoOf<T> = <<T as pallet_transaction_payment::Config>::OnChargeTransaction as OnChargeTransaction<T>>::LiquidityInfo;
+
+#[derive(Encode, Decode, DefaultNoBound)]
+pub enum InitialPayment<T: Config> {
+	Nothing,
+	Native(LiquidityInfoOf<T>),
+	Asset(CreditOf<T::AccountId, T::Fungibles>),
+}
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	use frame_support::{
+		dispatch::DispatchResultWithPostInfo,
+		pallet_prelude::*,
+		inherent::Vec,
+		traits::{
+			Currency, ReservableCurrency, EnsureOrigin, ExistenceRequirement::KeepAlive,
+		},
+		PalletId,
+	};
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_transaction_payment::Config + pallet_balances::Config + pallet_authorship::Config + pallet_assets::Config {
```

Comment:
```suggestion
	pub trait Config: frame_system::Config + pallet_transaction_payment::Config + pallet_assets::Config {
```

pfff, do we need all of this?

Reply:
Not sure we will need them in the end, just added them for easy hacking.
Authorship will probably be necessary because we access the block author to pay fees to them.
Balances we can most likely remove.

Reply:
There is a clear way to listen to authorship events: you implement `pallet_authorship::EventHandler` and add yourself to the list of authorship event handlers. 

Still not pretty as you need to import `pallet_authorship`, but your config need not be a sub-trait of it.

Reply:
And then I store the author? Because I need access to the author on when the transaction payment is triggered.

Reply:
I won't block on this, but the clean thing to do is to, similar to `CurrencyAdapter`, just accept a `T: OnUnBalanced` and then give it to a top level `struct DealWithFees` that can read the author or whatever.